### PR TITLE
1084 qiime map

### PR DESCRIPTION
**Pull request by josenavas**
Built on top of #1207 

This PR adds the ability to upload a QIIME mapping file trough the interface. When uploading a QIIME mapping file, a SampleTemplate and a PrepTemplate is created.

### CODE REVIEW

**Comment by antgonza** [comment]
Works fine. Except that the validation doesn't accept comments (lines that start with #). Also, I know is not part of this issue but it will be great if you could fix the warnings when adding a template. The issue is that if you have warnings about EBI, processing, etc all of them are in a single line, could you put them in different ones (add `</br>`)?  Gonna take a look a the code.

File: qiita_db/metadata_template/util.py [hunk 1]
```diff
@@ -315,3 +320,38 @@ def get_invalid_sample_names(sample_names):
...
+    some other different column.
+    """
+    first_line = None
+    with open_file(fp, mode='U') as f:
+        first_line = f.readline()
+    if not first_line:
+        raise QiitaDBError('Empty file passed!')
+
+    first_col = first_line.split()[0]
+    return first_col == '#SampleID'
```
Comment (antgonza) [inline qiita_db/metadata_template/util.py]: Should you also check for LinkerPrimer, BarcodeSequence, and ReverseBarcodeSequence (optional)?

File: qiita_db/metadata_template/test/test_util.py [hunk 2]
```diff
@@ -685,6 +719,19 @@ def test_invalid_lat_long(self):
...
      'str_column': {'2.Sample1': 'NA', '2.Sample2': 'NA', '2.Sample3': 'NA'}}
 
+QIIME_TUTORIAL_MAP_DICT_FORM = {
+    'BarcodeSequence': {'PC.354': 'AGCACGAGCCTA',
+                        'PC.607': 'AACTGTGCGTAC'},
+    'LinkerPrimerSequence': {'PC.354': 'YATGCTGCCTCCCGTAGGAGT',
+                             'PC.607': 'YATGCTGCCTCCCGTAGGAGT'},
+    'Treatment': {'PC.354': 'Control',
+                  'PC.607': 'Fast'},
+    'DOB': {'PC.354': 20061218,
```
Comment (ElDeveloper) [inline qiita_db/metadata_template/test/test_util.py]: Strictly speaking, QIIME mapping files do not do any sort of data type inferences, but I guess it's OK in this context as we need these data types to insert the information into the DB?

**Comment by antgonza** [comment]
Code looks fine, only one comment. However, not sure if we should have a qiime map loading reimplementation here or use qiime. I think having it here could help us deprecate the other one for qiime 2.

File: qiita_pet/handlers/study_handlers/description_handlers.py [hunk 3]
```diff
@@ -177,25 +178,33 @@ def process_sample_template(self, study, user, callback):
             # The file does not exist, fail nicely
             raise HTTPError(404, "This file doesn't exist: %s" % fp_rsp)
 
+        # Define here the message and message level in case of success
```
Comment (ElDeveloper) [inline qiita_pet/handlers/study_handlers/description_handlers.py]: this comment is outdated.

File: qiita_pet/templates/study_description_templates/study_information_tab.html [hunk 4]
```diff
@@ -12,15 +12,25 @@
   <hr>
 
   {% if show_select_sample %}
-    Select your sample template <br/>
+    Select your sample template or, alternatively, a QIIME mapping file <br/>
     (only files with the "txt" and "tsv" file extensions will be displayed here):
     <select id="sample_template">
       {% for f in files %}
-        {% if f.endswith(('txt', 'tsv')) %}
```
Comment (ElDeveloper) [inline qiita_pet/templates/study_description_templates/study_information_tab.html]: Why did you remove this check?

File: qiita_ware/test/test_metadata_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        for fp in self._clean_up_files:
+            if exists(fp):
+                remove(fp)
+
+    def test_create_templates_from_qiime_mapping_file(self):
+        new_pt_id = get_count('qiita.prep_template') + 1
+        obs_st, obs_pt = create_templates_from_qiime_mapping_file(
+            StringIO(QIIME_MAP), self.new_study, "16S")
+
+        # Be green: clean the environment
```
Comment (ElDeveloper) [inline qiita_ware/test/test_metadata_pipeline.py]: :sunglasses:

File: qiita_ware/metadata_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    -------
+    (SampleTemplate, PrepTemplate)
+        The templates created from the QIIME mapping file
+    """
+    qiime_map = load_template_to_dataframe(fp, index='#SampleID')
+
+    # There are a few columns in the QIIME mapping file that are special and
+    # we know how to deal with them
+    rename_cols = {
+        'BarcodeSequence': 'barcode',
```
Comment (ElDeveloper) [inline qiita_ware/metadata_pipeline.py]: What happens when none of these columns are present? Would an empty prep template be generated? Should we error in this case?

**Comment by ElDeveloper** [comment]
Code looks great, I'm gonna test the system now.

@biocore/qiita-admin would be good if others could test this.

**Comment by ElDeveloper** [comment]
As for having two parsers, I almost want to re-use the code in QIIME,
having multiple parsers for the same file-format is a nightmare and
prone to error. What do you guys think, I know there's the whole GPL
limitation, etc ... but then again, I think we should think this
through.

**Comment by josenavas** [comment]
do you want to have a call about the 2 parsers thing? I think it will be faster than discussing over here.

**Comment by ElDeveloper** [comment]
Sure, anyone else who would like to join?

File: qiita_db/metadata_template/util.py (hunk 1 above)
Comment (josenavas) [inline qiita_db/metadata_template/util.py]: This is just a quick and fast check to see if the file is a mapping file. Note that templates do not have any # sign at the beginning, not even for comments. The fact that the file starts with '#SampleID' is enough to know that we are not parsing a template, but a mapping file.

File: qiita_db/metadata_template/test/test_util.py (hunk 2 above)
Comment (josenavas) [inline qiita_db/metadata_template/test/test_util.py]: Yes, we need the type inference to be able to search correctly over metadata. Also, the idea is that the DB will not know that this data came from a mapping file.

File: qiita_pet/handlers/study_handlers/description_handlers.py (hunk 3 above)
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/description_handlers.py]: It looks like I end up removing two lines that I shouldn't. I've re-added the lines that define the message and message level in case of success.

File: qiita_pet/templates/study_description_templates/study_information_tab.html (hunk 4 above)
Comment (josenavas) [inline qiita_pet/templates/study_description_templates/study_information_tab.html]: I move this check to the handler, so less python is executed on the template.

File: qiita_ware/metadata_pipeline.py (hunk 6 above)
Comment (josenavas) [inline qiita_ware/metadata_pipeline.py]: Good point. I was assuming "correct" QIIME mapping files, but it is possible that the users are not uploading this information. I've added a check.

**Comment by josenavas** [comment]
@ElDeveloper @antgonza this should be ready to review!

File: qiita_pet/templates/study_description_templates/study_information_tab.html (hunk 4 above)
Comment (adamrp) [inline qiita_pet/templates/study_description_templates/study_information_tab.html]: :+1:

File: qiita_pet/handlers/study_handlers/description_handlers.py [hunk 7]
```diff
@@ -177,25 +178,35 @@ def process_sample_template(self, study, user, callback):
...
+        msg = "The sample template '%s' has been added" % sample_template
+        msg_level = "success"
+        is_mapping_file = looks_like_qiime_mapping_file(fp_rsp)
```
Comment (adamrp) [inline qiita_pet/handlers/study_handlers/description_handlers.py]: I like the escalation from "looks like" to "is" :laughing:

File: qiita_db/metadata_template/util.py [hunk 8]
```diff
@@ -315,3 +331,115 @@ def get_invalid_sample_names(sample_names):
...
+    This is not doing a validation of the QIIME mapping file. It simply checks
+    the first line in the file and it returns true if the line starts with
+    '#SampleID', since a sample/prep template will start with 'sample_name' or
+    some other different column.
+    """
+    first_line = None
+    with open_file(fp, mode='U') as f:
+        first_line = f.readline()
+    if not first_line:
+        raise QiitaDBError('Empty file passed!')
```
Comment (adamrp) [inline qiita_db/metadata_template/util.py]: Why raise an error here instead of just returning `False`?

File: qiita_pet/handlers/study_handlers/description_handlers.py (hunk 7 above)
Comment (josenavas) [inline qiita_pet/handlers/study_handlers/description_handlers.py]: hahaha, yeah it was way shorter....

File: qiita_db/metadata_template/util.py (hunk 8 above)
Comment (josenavas) [inline qiita_db/metadata_template/util.py]: That is actually a very good point. An empty file does not look like a QIIME mapping file, so the right thing to do is return False. The process will raise in another step though, but this will reduce the amount of code duplication that we have.

File: qiita_pet/templates/study_description_templates/study_information_tab.html (hunk 4 above)
Comment (ElDeveloper) [inline qiita_pet/templates/study_description_templates/study_information_tab.html]: :+1:

File: qiita_ware/metadata_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        'LinkerPrimerSequence': 'primer',
+        'Description': 'description',
+    }
+
+    if 'ReverseLinkerPrimer' in qiime_map:
+        rename_cols['ReverseLinkerPrimer'] = 'reverselinkerprimer'
+
+    missing = set(rename_cols).difference(qiime_map.columns)
+    if missing:
+        raise QiitaWareError(
```
Comment (ElDeveloper) [inline qiita_ware/metadata_pipeline.py]: Can you add this exception to this [try/except](https://github.com/biocore/qiita/blob/master/qiita_pet/handlers/study_handlers/description_handlers.py#L192).

**Comment by ElDeveloper** [comment]
PS - You also have some PEP8 failures:

```
qiita_db/metadata_template/util.py:400:13: E731 do not assign a lambda expression, use a def
qiita_db/metadata_template/util.py:403:13: E731 do not assign a lambda expression, use a def
qiita_db/metadata_template/util.py:407:13: E731 do not assign a lambda expression, use a def
qiita_db/metadata_template/util.py:410:13: E731 do not assign a lambda expression, use a def
```

File: qiita_ware/metadata_pipeline.py (hunk 9 above)
Comment (josenavas) [inline qiita_ware/metadata_pipeline.py]: Done

**Comment by josenavas** [comment]
Thanks @ElDeveloper . I don't understand why the flake8 in travis is different from the one on my machine... I think the one in conda has something different.

**Comment by ElDeveloper** [comment]
Thanks @josenavas. Yeah, conda is sometimes a bit outdated compared to the pypi packages ...

**Comment by ElDeveloper** [comment]
I hereby officially declare this as :+1:! If someone else wants to go ahead and merge, that would be awesome.

**Comment by antgonza** [comment]
```bash
In nomine Patris, et Filii, et Spiritus Sancti. Amen
```
